File: packages/paso-py/usepaso/utils/coerce.py
import re

_INTEGER_RE = re.compile(r'^-?\d+$')
_NUMBER_RE = re.compile(r'^-?\d+(\.\d+)?$')


def coerce_value(raw: str, declared_type: str, key: str):
    """Coerce a CLI string value to the declared input type. Raises ValueError on invalid input."""
    if declared_type == 'integer':
        if not _INTEGER_RE.match(raw):
            raise ValueError(f'Parameter "{key}" must be an integer, got "{raw}"')
        return int(raw)
    elif declared_type == 'number':
        if not _NUMBER_RE.match(raw):
            raise ValueError(f'Parameter "{key}" must be a number, got "{raw}"')
        return float(raw)
    elif declared_type == 'boolean':
        if raw == 'true':
            return True
        elif raw == 'false':
            return False
        raise ValueError(f'Parameter "{key}" must be true or false, got "{raw}"')
    else:
        return raw

File: packages/paso-js/src/utils/coerce.ts
const INTEGER_RE = /^-?\d+$/;
const NUMBER_RE = /^-?\d+(\.\d+)?$/;

export function coerceValue(raw: string, type: string, key: string): unknown {
  switch (type) {
    case 'integer': {
      if (!INTEGER_RE.test(raw)) {
        throw new Error(`Parameter "${key}" must be an integer, got "${raw}"`);
      }
      return parseInt(raw, 10);
    }
    case 'number': {
      if (!NUMBER_RE.test(raw)) {
        throw new Error(`Parameter "${key}" must be a number, got "${raw}"`);
      }
      return parseFloat(raw);
    }
    case 'boolean':
      if (raw === 'true') return true;
      if (raw === 'false') return false;
      throw new Error(`Parameter "${key}" must be true or false, got "${raw}"`);
    case 'string':
    case 'enum':
      return raw;
    default:
      return raw;
  }
}
